src/EventStore.Core.Tests/ClientAPI/catchup_filtered_subscription.cs

```diff
@@ -0,0 +1,244 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using EventStore.Core.Services;
+
+namespace EventStore.Core.Tests.ClientAPI {
+	[TestFixture]
+	[Category("LongRunning"), Category("ClientAPI")]
+	public class catchup_filtered_subscription : SpecificationWithDirectory {
+		private MiniNode _node;
+		private IEventStoreConnection _conn;
+		private List<EventData> _testEvents;
+		private List<EventData> _testEventsAfter;
+		private const int Timeout = 10000;
+
+		[SetUp]
+		public override void SetUp() {
+			base.SetUp();
+			_node = new MiniNode(PathName);
+			_node.Start();
+
+			_conn = BuildConnection(_node);
+			_conn.ConnectAsync().Wait();
+			_conn.SetStreamMetadataAsync("$all", -1,
+				StreamMetadata.Build().SetReadRole(SystemRoles.All),
+				new UserCredentials(SystemUsers.Admin, SystemUsers.DefaultAdminPassword)).Wait();
+
+			_testEvents = Enumerable
+				.Range(0, 10)
+				.Select(x => TestEvent.NewTestEvent(x.ToString(), eventName: "AEvent"))
+				.ToList();
+
+			_testEvents.AddRange(
+				Enumerable
+					.Range(0, 10)
+					.Select(x => TestEvent.NewTestEvent(x.ToString(), eventName: "BEvent")).ToList());
+
+			_testEventsAfter = Enumerable
+				.Range(0, 10)
+				.Select(x => TestEvent.NewTestEvent(x.ToString(), eventName: "AEvent"))
+				.ToList();
+
+			_testEventsAfter.AddRange(
+				Enumerable
+					.Range(0, 10)
+					.Select(x => TestEvent.NewTestEvent(x.ToString(), eventName: "BEvent")).ToList());
+
+			_conn.AppendToStreamAsync("stream-a", ExpectedVersion.NoStream, _testEvents.EvenEvents()).Wait();
+			_conn.AppendToStreamAsync("stream-b", ExpectedVersion.NoStream, _testEvents.OddEvents()).Wait();
+		}
+
+		protected virtual IEventStoreConnection BuildConnection(MiniNode node) {
+			return TestConnection.Create(node.TcpEndPoint);
+		}
+
+		[Test]
+		public void calls_checkpoint_delegate_during_catchup() {
+			var filter = Filter.StreamId.Prefix("stream-a");
+			var appeared = new CountdownEvent(9);
+			var eventsSeen = 0;
+
+			var settings = new CatchUpSubscriptionFilteredSettings(
+				10000,
+				2,
+				verboseLogging: false,
+				resolveLinkTos: true,
+				maxSearchWindow: 2,
+				subscriptionName: String.Empty
+			);
+
+			_conn.FilteredSubscribeToAllFrom(
+				Position.Start,
+				filter,
+				settings,
+				(s, e) => {
+					Console.WriteLine("e");
+					eventsSeen++;
+					return Task.CompletedTask;
+				},
+				(s, p) => {
+					Console.WriteLine("x");
+					appeared.Signal();
+					return Task.CompletedTask;
+				}, 1);
+
+			if (!appeared.Wait(Timeout)) {
+				Assert.Fail("Checkpoint appeared not called enough times within time limit.");
+			}
+
+			Assert.AreEqual(10, eventsSeen);
+		}
+
+		[Test]
+		public void calls_checkpoint_during_live_processing_stage() {
+			var filter = Filter.StreamId.Prefix("stream-a");
+			var appeared = new CountdownEvent(6);
```

Comment:
I wonder if we can't be a little more explicit here about what the 6 here represents... 5 x checkpoint reached + 1 x live processing started checkpoint reached called.